### DISCUSSION TITLE
Alternative or workarounds for Helm's .Release.isUpgrade?

### DISCUSSION THREAD

**Discussion by jgoeres** [argoproj/argo-cd#7496] [Q&A]
We are setting up our SaaS product with ArgoCD using a Helm chart. 
As part of the initial deployment, we run a few K8s jobs. To make sure that they run only initially, the job templates are guarded with an "not if .Release.isUpgrade". 
Alas, this Helm flag is never set by ArgoCD (I am aware that there is a statement about that in the docs at https://argo-cd.readthedocs.io/en/stable/user-guide/helm/#helm-hooks ).
This can now lead to the situation that we make changes to the configuration and run a sync that also affects the job spec - one common example is changing the image tag. We have a common setting for the tag that is by default used for all our container images, _including_ the jobs (but can be overridden for individual services to rollout patches). 
What now happens if we roll out an update is that ArgoCD will attempt to also change the image version of the job spec, leading to a "field is immutable" error and causing the entire deployment to end up (and stay!) "out of sync", see screnshots below.
We are currently using v2.0.3.

Are there any plans to address these and similar issues? Like making ArgoCD aware of whether a sync is an initial helm install or a helm upgrade? Or does anyone have a proposal for working around this issue?

![image](https://user-images.githubusercontent.com/1484812/137275029-98ec3fdc-e293-45da-ad38-b661ec9cf1ca.png)

![image](https://user-images.githubusercontent.com/1484812/137275848-53802bc9-2c81-4342-a298-fb5791781c58.png)

**Comment by hawksight** [comment]
I too was hoping to use this flag at one point whilst trying to build in applications migrations to our helm chart via the helm hooks. The key for me was that the migrations job ended up need all the env vars, configmaps and secrets that the main deployment had. So on a new install I was going to execute migrations post sync, or post install. Theory being, I don't need to pre-create secrets and configmap just for the job, and that on first install, the app didn't need to work instantly, if a lack of migrations broke service. It would be acceptable for things to come alive after the job had run.

But on an upgrade we actually want the job to run before sync. In this scenario the secrets and configmaps already exist, so in that case, I don't need to pre-create extra resources. So this combination of events was the driver for me using this helm built in variable `.Release.isUpgrade`.

However as you say, it'd not available in ArgoCD world, so supporting it would actually be every helpful.

**Comment by jessesuen** [comment]
Unlike helm, Argo CD does not keep track of helm lifecycle states (e.g. to store history and rollback information). Helm accomplishes this by storing lifecycle information in a Secret. See: https://helm.sh/docs/topics/advanced/#storage-backends

Argo CD fundamentally works by running `kubectl apply` on resources. We must do it this way because we rely on kubectl's `last-applied-configuration` annotation set by kubectl apply in order to properly calculate diffs.

So in short, our Helm support is similar to:  `helm template <options> | kubectl apply -f -`. Because of this, we do not really understand when an application is "upgrading" vs. "installing," unlike Helm which consults its storage backend during the install.

The general guidance we have about `Release.isUpgrade` is to instead use the default `isInstall` but to make the desired logic idempotent and safe to re-apply / re-run.

**Comment by jgoeres** [comment]
What could help to somewhat work around this issue is if ArgoCD would set some custom Helm properties to which the Helmchart could react. That would make it possible to use the regular Helm hooks when the Helm chart is used with Helm directly, and behave differently if running within ArgoCD, e.g.

    {{ if ArgoCd.isArgoCd }}
      #Stuff done when running in ArgoCD
      ...
    {{ else }}
      #different stuff done when NOT running in ArgoCD
      ...   

Thoughts?

**Reply by hawksight** [reply]
ArgoCD injects a label to track application resource.. so maybe in the chart, you could have a helper function to work out if it's being deployed with ArgoCD or not. Something like...

```
{{- define "chart.isArgocd" -}}
{{- if "argocd.argoproj.io/instance" in .Values.labels }}
true
{{- else -}}
false
{{- end -}}
{{- end -}}
```

^ Don't think that works but couldn't find a quick example, so treat that as pseudo code
